### PR TITLE
feat(bedrock): implement new data source structure

### LINKED ISSUE #655: (bedrock): support additional data sources
### Describe the feature

New data sources config are available in cdk v2.154.1
https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_bedrock.CfnDataSource.DataSourceConfigurationProperty.html 

### Use Case

Missing features

### Proposed Solution

_No response_

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by aws-rafams**
Fixes #655

Draft, work in progress.

Just linking the draft to get some comments on the planned structure / interface.

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the project license.

### CODE REVIEW

File: src/cdk-lib/bedrock/data-sources/base-data-source.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+   * @example 'JHUEVXUZMU'
+   */
+  readonly dataSourceId: string;
+}
+
+/**
+ * Specifies the base class for all data source resources (imported and new).
+ * @abstract
+ */
+export abstract class DataSourceBase extends Resource implements IDataSource {
```
Comment (krokoko) [inline src/cdk-lib/bedrock/data-sources/base-data-source.ts:57]: Thanks for the new structure, it is better :) Follows CDK's guidance

**Comment by aws-rafams** [comment]
The implementation is practically finished, just needs extensive testing.  Would love to have comments on the final structure and on the interfaces. Once the structure has been validated, will proceed with:
- further unit tests
- further integ tests
- documentation updates

Working example
```ts
const kb = new KnowledgeBase(stack, 'MyKnowledgeBase', {
  name: 'MyKnowledgeBase',
  embeddingsModel: BedrockFoundationModel.COHERE_EMBED_MULTILINGUAL_V3,
});

const bucket = new Bucket(stack, 'Bucket', {});
const lambdaFunction = new Function(stack, 'MyFunction', {
  runtime: cdk.aws_lambda.Runtime.PYTHON_3_9,
  handler: 'index.handler',
  code: cdk.aws_lambda.Code.fromInline('print("Hello, World!")'),
});

kb.addWebCrawlerDataSource({
  sourceUrls: ['https://docs.aws.amazon.com/'],
  chunkingStrategy: ChunkingStrategy.HIERARCHICAL_COHERE,
  customTransformation: CustomTransformation.lambda({
    lambdaFunction: lambdaFunction,
    s3BucketUri: `s3://${bucket.bucketName}/chunk-processor/`,
  }),
});

kb.addS3DataSource({
  bucket,
  chunkingStrategy: ChunkingStrategy.SEMANTIC,
  parsingStrategy: ParsingStategy.foundationModel({
    parsingModel: BedrockFoundationModel.ANTHROPIC_CLAUDE_SONNET_V1_0.asIModel(stack),
  }),
});
```

**Comment by krokoko** [comment]
Just FYI CDK v[.2.155](https://github.com/aws/aws-cdk/releases/tag/v2.155.0) has updates impacting the resources used in this PR, see https://github.com/aws/aws-cdk/pull/31193

File: src/cdk-lib/bedrock/data-sources/parsing.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+   * @see https://docs.aws.amazon.com/bedrock/latest/userguide/knowledge-base-ds.html#kb-ds-supported-doc-formats-limits
+   */
+  public static foundationModel(props: FoundationModelParsingStategyProps): ParsingStategy {
+    return {
+      configuration: {
+        bedrockFoundationModelConfiguration: {
+          modelArn: props.parsingModel.modelArn,
+          parsingPrompt: {
+            parsingPromptText: props.parsingPrompt
+              ?? fs.readFileSync(path.join(__dirname, 'default-parsing-prompt.txt'), 'utf8'),
```
Comment (krokoko) [inline src/cdk-lib/bedrock/data-sources/parsing.ts:70]: do we need a default prompt if not provided ?

**Comment by krokoko** [comment]
Overall LGTM ! Thanks for this ! @aws-rafams I am fine with the structure and interfaces, do you still need to work on: 
- further unit tests
- further integ tests
- documentation updates
?
If not, I will run some tests with the package through our samples
There is couple of breaking changes too that we will need to highlight in the release notes

File: src/cdk-lib/bedrock/data-sources/parsing.ts (hunk 2 above)
Comment (aws-rafams) [reply src/cdk-lib/bedrock/data-sources/parsing.ts]: I think this is solved with commit [62fc779](https://github.com/awslabs/generative-ai-cdk-constructs/pull/668/commits/62fc7793555208f20221bd2109e6335cdc82ead9)

**Comment by aws-rafams** [comment]
> Overall LGTM ! Thanks for this ! @aws-rafams I am fine with the structure and interfaces, do you still need to work on:
> 
> * further unit tests
> * further integ tests
> * documentation updates
>   ?
>   If not, I will run some tests with the package through our samples
>   There is couple of breaking changes too that we will need to highlight in the release notes

I have completed the tests and documentation updates. However, I would really appreciate a hand with the Python snippets on the docs, which I believe are the only remaining task. 

The primary breaking change is in the S3DataSource resource and the ChunkingStrategy enum which is now a class. The previous structure and properties will not work with the new version, so we should highlight this in the release notes. 

The changes to the KnowledgeBase are primarily the addition of helper methods, while the rest are new resources that did not exist in the previous version.

**Comment by krokoko** [comment]
> > Overall LGTM ! Thanks for this ! @aws-rafams I am fine with the structure and interfaces, do you still need to work on:
> > 
> > * further unit tests
> > * further integ tests
> > * documentation updates
> >   ?
> >   If not, I will run some tests with the package through our samples
> >   There is couple of breaking changes too that we will need to highlight in the release notes
> 
> I have completed the tests and documentation updates. However, I would really appreciate a hand with the Python snippets on the docs, which I believe are the only remaining task.
> 
> The primary breaking change is in the S3DataSource resource and the ChunkingStrategy enum which is now a class. The previous structure and properties will not work with the new version, so we should highlight this in the release notes.
> 
> The changes to the KnowledgeBase are primarily the addition of helper methods, while the rest are new resources that did not exist previously.

Thanks @aws-rafams, I'll run some tests with the package and update here !

**Comment by krokoko** [comment]
Thanks @aws-rafams !
Ran several tests using our bedrock agent sample, everything seems to work as expected. Non exhaustive list includes:

- testing the different chunking strategies
- import of KB

image

As discussed, only detected issue is on permissions related to FM parsing and lambda in custom transformation that are missing. Readme seems good, we will update it soon as we need to add support for the new languages.

Could you please:
- Fix the permission issue
- Remove the integ tests causing build failure
After that we can approve and merge

Note:
- We can add the changes introduced in cdk v2.155 (see https://github.com/awslabs/generative-ai-cdk-constructs/pull/668#issuecomment-2329770111) through a different PR

**Comment by aws-rafams** [comment]
The permission issue has been solved: 
Screenshot 2024-09-25 at 14 42 31

**Review by krokoko — APPROVED** [review]

**Pull request merged**